### ISSUE TITLE
panic when YAML key is a YAML bool

### ISSUE THREAD

**Issue by jbuck** [mozilla/sops#194]
While running `sops -d filename` across the cloud-ops secrets repo it crashed when attempting to decrypt msisdn files.

panic: interface conversion: interface is bool, not string

goroutine 1 [running]:
panic(0x49a160, 0xc42028c800)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
go.mozilla.org/sops.TreeBranch.walkBranch(0xc420585700, 0x7, 0x8, 0xc420585800, 0x5, 0x8, 0xc420436ea0, 0x1, 0x1, 0xc4205b8f48, ...)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/sops.go:185 +0xca
go.mozilla.org/sops.TreeBranch.walkValue(0xc420585700, 0x7, 0x8, 0x4baea0, 0xc4203fb900, 0xc420436ea0, 0x1, 0x1, 0xc4205b8f48, 0x4baea0, ...)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/sops.go:158 +0x61e
go.mozilla.org/sops.TreeBranch.walkBranch(0xc420585700, 0x7, 0x8, 0xc420585700, 0x7, 0x8, 0x785f68, 0x0, 0x0, 0xc4205b8f48, ...)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/sops.go:185 +0x196
go.mozilla.org/sops.Tree.Decrypt(0xc420585700, 0x7, 0x8, 0xecea0b9fa, 0x0, 0x765200, 0xc420444560, 0xc, 0xc4205a2c60, 0x117, ...)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/sops.go:261 +0x1c9
main.decryptTree(0xc420585700, 0x7, 0x8, 0xecea0b9fa, 0x0, 0x765200, 0xc420444560, 0xc, 0xc4205a2c60, 0x117, ...)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/cmd/sops/main.go:344 +0x2dd
main.decrypt(0xc4200ab180, 0xc420585700, 0x7, 0x8, 0xecea0b9fa, 0x0, 0x765200, 0xc420444560, 0xc, 0xc4205a2c60, ...)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/cmd/sops/main.go:356 +0x92
main.main.func1(0xc4200ab180, 0x0, 0x0)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/cmd/sops/main.go:248 +0x5da
go.mozilla.org/sops/vendor/gopkg.in/urfave/cli%2ev1.HandleAction(0x47aae0, 0x55bb90, 0xc4200ab180, 0x0, 0x0)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/vendor/gopkg.in/urfave/cli.v1/app.go:485 +0xd4
go.mozilla.org/sops/vendor/gopkg.in/urfave/cli%2ev1.(*App).Run(0xc420127040, 0xc420078090, 0x3, 0x3, 0x0, 0x0)
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/vendor/gopkg.in/urfave/cli.v1/app.go:259 +0x74f
main.main()
	/Users/jbuckley/Documents/go/src/go.mozilla.org/sops/cmd/sops/main.go:284 +0xe14

**Comment by autrilla** [comment]
This is due to boolean keys in YAML documents. So something like:

```
True: ENC[...]
```
Does not decrypt.

**Comment by autrilla** [comment]
This may be tricky to fix because the Python YAML implementations are off-spec:

```
>>> import ruamel.yaml
>>> from pprint import pprint
>>> bools = "y|Y|yes|Yes|YES|n|N|no|No|NO|true|True|TRUE|false|False|FALSE|on|On|ON|off|Off|OFF".split("|") # boolean values according to the YAML spec
>>> bools = dict(zip(bools, [ruamel.yaml.load(b + ": 1") for b in bools]))
>>> pprint(bools)
{'FALSE': {False: 1},
 'False': {False: 1},
 'N': {'N': 1},
 'NO': {False: 1},
 'No': {False: 1},
 'OFF': {False: 1},
 'ON': {True: 1},
 'Off': {False: 1},
 'On': {True: 1},
 'TRUE': {True: 1},
 'True': {True: 1},
 'Y': {'Y': 1},
 'YES': {True: 1},
 'Yes': {True: 1},
 'false': {False: 1},
 'n': {'n': 1},
 'no': {False: 1},
 'off': {False: 1},
 'on': {True: 1},
 'true': {True: 1},
 'y': {'y': 1},
 'yes': {True: 1}}
```

Whereas with Go:

```
y map[true:1]
Y map[true:1]
yes map[true:1]
Yes map[true:1]
YES map[true:1]
n map[false:1]
N map[false:1]
no map[false:1]
No map[false:1]
NO map[false:1]
true map[true:1]
True map[true:1]
TRUE map[true:1]
false map[false:1]
False map[false:1]
FALSE map[false:1]
on map[true:1]
On map[true:1]
ON map[true:1]
off map[false:1]
Off map[false:1]
OFF map[false:1]
```

So files encrypted with the Python version that contain `y` or `n` as YAML keys will fail the MAC check. For the rest, the Go version should simply use the strings `True` or `False` for the MAC whenever a boolean is encountered.

**Comment by autrilla** [comment]
Looking into this further, python-sops does require keys to be strings. Because of the non-compliance of libyaml with the YAML spec, it didn't consider `y`, `Y`, `n` or `N` a boolean, so they were allowed as keys.

I think the way to proceed is to leave it as it is, allowing only string keys. Provide an error message for this instead of panicking. Files with keys that should actually be booleans but weren't considered such will need to have those keys changed.

cc @jvehent

**Comment by jvehent** [comment]
That seems fine. It is an edge cae (booleans as keys? really?) and handling it with an error instructing users to reset the MAC seems like the right way to solve it.

**Comment by autrilla** [comment]
Fixed by #197
